Tasks/VsTest/vstest.ts

```diff
@@ -1377,63 +1386,61 @@ function resetRunInParallel() {
 }
 
 
-function getLatestVSTestConsolePathFromRegistry(): Q.Promise<[number, string]> {
-    let deferred = Q.defer<[number,string]>();
+function getLatestVSTestConsolePathFromRegistry(): Q.Promise<VSTestConsoleInfo> {
+    let deferred = Q.defer<VSTestConsoleInfo>();
     var regPath = 'HKLM\\SOFTWARE\\Microsoft\\VisualStudio';
     regedit.list(regPath).on('data', (entry) => {
         let subkeys = entry.data.keys;
         let versions = getFloatsFromStringArray(subkeys);
         if (versions && versions.length > 0) {
             versions.sort((a, b) => a-b);
             let selectedVersion = versions[versions.length - 1];
-            deferred.resolve([selectedVersion, getVSTestLocation(selectedVersion)]);
+            deferred.resolve(new VSTestConsoleInfo(selectedVersion, getVSTestLocation(selectedVersion)));
             return deferred.promise;
         }
     });
     deferred.resolve(null);
     return deferred.promise;
 }
 
-function getVSTestConsole15Path(): Q.Promise<string> {
+function getVSTestConsole15Path(): string {
     let powershellTool = tl.tool('powershell');
     let powershellArgs = ['-file', path.join(__dirname, 'vs15Helper.ps1')]
     powershellTool.arg(powershellArgs);
     let xml = powershellTool.execSync().stdout;
     let deferred = Q.defer<string>();
-    Q.nfcall(xml2js.parseString, xml).then((data) => {
-        try {
-            let vs15InstallDir = data['Objs']['S'][0];
-            let vstestconsolePath =  path.join(vs15InstallDir, 'Common7', 'IDE', 'CommonExtensions', 'Microsoft', 'TestWindow', 'vstest.console.exe');
-            deferred.resolve(vstestconsolePath);
-        } catch (e) {
-            deferred.resolve(null);
+    let vstestconsolePath:string = null;
+    xml2js.parseString(xml, (err, result) => {
+        if(result) {
+            try {
+                let vs15InstallDir = result['Objs']['S'][0];
+                vstestconsolePath = path.join(vs15InstallDir, 'Common7', 'IDE', 'CommonExtensions', 'Microsoft', 'TestWindow', 'vstest.console.exe');
+            } catch (e) {
+                vstestconsolePath = null;
+            }
         }
-    });
-    return deferred.promise;
+    })
+    return vstestconsolePath;
 }
 
 
-function locateVSVersion(): Q.Promise<[number, string]> {
-    var defer = Q.defer<[number, string]>();
-    let vsVersion: number = parseFloat(vsTestVersion);
+function locateVSVersion(version: string): Q.Promise<VSTestConsoleInfo> {
+    let deferred = Q.defer<VSTestConsoleInfo>();
```

Comment:
get rid of these deferred vars. your promises can be chained, can be returned directly. code is cleaner and simpler